Matchismo/CardMatchingGame.m

```diff
@@ -6,163 +6,130 @@
+        [self setIsGameMode3Way:NO];
+        self.cards = [NSMutableArray array];
+        for (int i = 0; i < count; i++) {
             Card *card = [deck drawRandomCard];
-            if (card){
+            if (card) {
                 [self.cards addObject:card];
             }
+            else {
+                return nil;
+            }
         }
     }
     return self;
 }
 
-- (instancetype)init{
+- (instancetype)init {
     return nil;
 }
 
 
--(void)setGameMode:(BOOL)gameModeIs3Way
-{
-    if (gameModeIs3Way)
-    {
+#pragma mark - Public API
+
+- (Card *)cardAtIndex:(NSUInteger)index {
+    return index < [self.cards count] ? self.cards[index] : nil;
+}
+
+- (void)setIsGameMode3Way:(BOOL)gameModeIs3Way {
+    if (gameModeIs3Way) {
         [self set3WayMatchScoring];
-        self.currentAllowedChosen = ALLOWED_CHOSEN_3WAY;
+        self.currentAllowedChosen = kSetAllowedChosen;
     } else {
         [self set2WayMatchScoring];
-        self.currentAllowedChosen = ALLOWED_CHOSEN_2WAY;
+        self.currentAllowedChosen = kMatchismoAllowedChosen;
     }
 }
 
-- (void)set2WayMatchScoring
-{
-    self.revealCost = 1;
-    self.mismatchPenalty = 2;
-    self.matchBonus = 4;
-}
-- (void)set3WayMatchScoring
-{
-    self.revealCost = 1;
-    self.mismatchPenalty = 3;
-    self.matchBonus = 4;
-}
-
-- (Card *)cardAtIndex:(NSUInteger)index
-{
-    return index < [self.cards count] ? self.cards[index] : nil;
-}
-
-- (void)chooseCardAtIndex:(NSUInteger)index
-{
+- (void)chooseCardAtIndex:(NSUInteger)index {
     Card *card = [self cardAtIndex:index];
-    if (card.matched)
-    {
+    if (card.matched) {
         return;
     }
     self.lastMatchedCards = nil;
-    if (card.chosen)
-    {
+    if (card.chosen) {
         card.chosen = NO;
-    }
-    else
-    {
+    } else {
         [self handleCardReveal:card];
     }
 }
-- (void)handleCardReveal:(Card *)card
-{
-    NSMutableArray<Card *> *chosenCards = [self getChosenCards];
-    [chosenCards addObject:card];
-    if ([chosenCards count] == self.currentAllowedChosen)
-    {
-        [self handleMatch:chosenCards];
-    }
+
+
+#pragma mark - Game logic functions
+
+- (void)handleCardReveal:(Card *)card {
     self.score -= self.revealCost;
+    NSMutableArray<Card *> *chosenCards = [self getCurrentlyChosenCards];
+    if ([chosenCards count] + 1 == self.currentAllowedChosen) {
+        [self handleMatch:card withCards:chosenCards];
+    }
     card.chosen = YES;
 }
 
-- (NSMutableArray *)getChosenCards
-{
-    NSMutableArray *chosenCards = [[NSMutableArray alloc] init];
+- (NSMutableArray *)getCurrentlyChosenCards {
+    NSMutableArray *chosenCards = [NSMutableArray array];
     for (Card *otherCard in self.cards) {
-        if (!otherCard.matched && otherCard.chosen)
-        {
-            if ([chosenCards count] < self.currentAllowedChosen)
-            {
+        if (!otherCard.matched && otherCard.chosen) {
+            if ([chosenCards count] < self.currentAllowedChosen) {
```

Comment:
no reason for that to happen, as it would mean that 
```
if ([chosenCards count] + 1 == self.currentAllowedChosen) {
  [self handleMatch:card withCards:chosenCards];
}
```
Did not work well.
And if this is to protect for game mode changes - I'd suggest that a game mode change would reset the current selection.